pkg/liveshare/ports.go

```diff
@@ -0,0 +1,130 @@
+package liveshare
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/sourcegraph/jsonrpc2"
+	"golang.org/x/sync/errgroup"
+)
+
+// Port describes a port exposed by the container.
+type Port struct {
+	SourcePort                       int    `json:"sourcePort"`
+	DestinationPort                  int    `json:"destinationPort"`
+	SessionName                      string `json:"sessionName"`
+	StreamName                       string `json:"streamName"`
+	StreamCondition                  string `json:"streamCondition"`
+	BrowseURL                        string `json:"browseUrl"`
+	IsPublic                         bool   `json:"isPublic"`
+	IsTCPServerConnectionEstablished bool   `json:"isTCPServerConnectionEstablished"`
+	HasTLSHandshakePassed            bool   `json:"hasTLSHandshakePassed"`
+	Privacy                          string `json:"privacy"`
+}
+
+type PortChangeKind string
+
+const (
+	PortChangeKindStart  PortChangeKind = "start"
+	PortChangeKindUpdate PortChangeKind = "update"
+)
+
+type PortUpdate struct {
+	Port        int            `json:"port"`
+	ChangeKind  PortChangeKind `json:"changeKind"`
+	ErrorDetail string         `json:"errorDetail"`
+	StatusCode  int            `json:"statusCode"`
+}
+
+// startSharing tells the Live Share host to start sharing the specified port from the container.
+// The sessionName describes the purpose of the remote port or service.
+// It returns an identifier that can be used to open an SSH channel to the remote port.
+func (s *Session) startSharing(ctx context.Context, sessionName string, port int) (channelID, error) {
+	args := []interface{}{port, sessionName, fmt.Sprintf("http://localhost:%d", port)}
+	g, ctx := errgroup.WithContext(ctx)
+
+	g.Go(func() error {
+		startNotification, err := s.WaitForPortNotification(ctx, port, PortChangeKindStart)
+		if err != nil {
+			return fmt.Errorf("error while waiting for port notification: %w", err)
+
+		}
+		if !startNotification.Success {
+			return fmt.Errorf("error while starting port sharing: %s", startNotification.ErrorDetail)
+		}
+		return nil // success
+	})
+
+	var response Port
+	g.Go(func() error {
+		return s.rpc.do(ctx, "serverSharing.startSharing", args, &response)
+	})
+
+	if err := g.Wait(); err != nil {
+		return channelID{}, err
+	}
+
+	return channelID{response.StreamName, response.StreamCondition}, nil
+}
+
+type PortNotification struct {
+	PortUpdate
+	Success bool
+}
+
+// WaitForPortNotification waits for a port notification to be received. It returns the notification
+// or an error if the notification is not received before the context is cancelled or it fails
+// to parse the notification.
+func (s *Session) WaitForPortNotification(ctx context.Context, port int, notifType PortChangeKind) (*PortNotification, error) {
+	notificationUpdate := make(chan PortNotification, 1)
+	errc := make(chan error, 1)
+
+	h := func(success bool) func(*jsonrpc2.Conn, *jsonrpc2.Request) {
```

Comment:
I'm not sure what your intent is here, but the control logic here is strange. It could use comments at the very least.

Shouldn't the request handler be unregistered when the outer function returns? Otherwise it looks like it will continue to service JSON RPC request indefinitely, filling the errc and notification channels, and getting stuck.

It might be slightly simpler to use a single channel for each req.Params, and call unmarshal in the main goroutine.

Reply:
good point, I've made the change to deregister the handler once the outer function returns. To determine which handler function to remove from the handler function array, I compare the addresses of the handlers and if they're equal, the handler gets removed from the array. However, my knowledge of go is limited, so please check if this equality comparison is correct

Reply:
I think your intent is that the function returns either when the context is cancelled, or when the first matching notification occurs, or an error. 1-buffered channels are half the solution; the other half is to use non-blocking sends so that no goroutine gets stuck if a second event should occur:

```
	// We use 1-buffered channels and non-blocking sends so that
	// no goroutine gets stuck.
	notificationCh := make(chan *PortNotification, 1)
	errCh := make(chan error, 1)

	h := func(success bool) func(*jsonrpc2.Conn, *jsonrpc2.Request) {
		return func(conn *jsonrpc2.Conn, req *jsonrpc2.Request) {
			notification := new(PortNotification)
			if err := json.Unmarshal(req.Params, notification); err != nil {
				select {
				case errCh <- fmt.Errorf("error unmarshaling notification: %w", err):
				default:
				}
				return
			}
			notification.Success = success
			if notification.Port == port && notification.ChangeKind == notifType {
				select {
				case notificationCh <- notification:
				default:
				}
			}
		}
	}
	deregisterSuccess := s.registerRequestHandler("serverSharing.sharingSucceeded", h(true))
	deregisterFailure := s.registerRequestHandler("serverSharing.sharingFailed", h(false))
	defer deregisterSuccess()
	defer deregisterFailure()

	select {
	case <-ctx.Done():
		return nil, ctx.Err()
	case errCh := <-errc:
		return nil, err
	case n := <-notificationCh:
		return n, nil
	}
```

Reply:
makes sense, updated in latest commit